### PR TITLE
Fix whitelist inheritance

### PR DESCRIPTION
**Pull request by lubomudr**
Hi
In simple config
```
load_module     modules/ngx_http_naxsi_module.so;
error_log       log/error_debug.log debug;

events {
    worker_connections  128;
}

http {
    include             mime.types;
    default_type        application/octet-stream;

    include             naxsi/naxsi_core.rules;

    server  {
        listen localhost:8888;
        server_name naxsi;

        location /naxsi_test/       {
            SecRulesEnabled;
            DeniedUrl "/NaxsiDeniedAccess";
            BasicRule wl:1000 "mz:$URL_X:^/naxsi_test|URL";
            CheckRule "$SQL >= 8" BLOCK;

            location /naxsi_test/child/     {
                return 200 "200 $request_uri\n";
            }
            return 200 "200 $request_uri\n";
        }
        location /NaxsiDeniedAccess {
            return 403 "403 $request_uri\n";
        }
    }
}
```
Whitelist for rule 1000 does not inherit :-(
```
lubomudr@lvl-suse:~/nginx/test> curl  -H "Server: naxsi" --url 'http://localhost:8888/naxsi_test/update/table.gif'
200 /naxsi_test/update/table.gif

lubomudr@lvl-suse:~/nginx/test> curl  -H "Server: naxsi" --url 'http://localhost:8888/naxsi_test/child/update/table.gif'
403 /naxsi_test/child/update/table.gif
```
This happens in case of missing directives in the nested location and not placing the loc config into main_config->locations list in ngx_http_naxsi_merge_loc_conf. As a consequence, in ngx_http_naxsi_init function are not filled whitelist hash tables.

Perhaps it would be more correct to insert the location configuration into main_conf->locations immediately when calling ngx_http_naxsi_create_loc_conf , but when creating the first location configuration at the main level, ngx_http_conf_get_module_main_conf(cf, ngx_http_naxsi_module) returns NULL :-(

And cosmetically designed it as a separate routine

### CODE REVIEW

File: naxsi_src/naxsi_skeleton.c [hunk 1]
```diff
@@ -515,8 +535,12 @@ ngx_http_naxsi_merge_loc_conf(ngx_conf_t* cf, void* parent, void* child)
     conf->enabled = prev->enabled;
   if (conf->force_disabled == 0)
     conf->force_disabled = prev->force_disabled;
-  if (conf->pushed == 0)
-    conf->pushed = prev->pushed;
+
+//  if (conf->pushed == 0)
+//    conf->pushed = prev->pushed;
```
Comment (wargio) [inline naxsi_src/naxsi_skeleton.c:540]: don't comment it. just remove it

**Review by wargio — CHANGES_REQUESTED** [review]
please add a test.

File: naxsi_src/naxsi_skeleton.c (hunk 1 above)
Comment (lubomudr) [reply naxsi_src/naxsi_skeleton.c]: Ok

**Review by wargio — CHANGES_REQUESTED** [review]
Add a test under here: https://github.com/wargio/naxsi/blob/main/unit-tests/tests/05naxsi_advanced_whitelists.t#L414

Something like this.

```
=== WL TEST 8: Whitelists inheritance
--- user_files
>>> robots.txt
aaaa
--- main_config
load_module $TEST_NGINX_NAXSI_MODULE_SO;
--- http_config
include $TEST_NGINX_NAXSI_RULES;
MainRule id:20000003 "s:$UWA:8" "rx:\.txt$" "mz:URL" "msg:file access to .txt";
--- config
location / {
	BasicRule wl:20000003 "mz:$URL:/robots.txt|URL";
	#LearningMode;
	SecRulesEnabled;
	DeniedUrl "/RequestDenied";
	CheckRule "$SQL >= 8" BLOCK;
	CheckRule "$RFI >= 8" BLOCK;
	CheckRule "$TRAVERSAL >= 4" BLOCK;
	CheckRule "$XSS >= 8" BLOCK;
	location /child/     {
		return 200 "200 $request_uri";
	}
  	root $TEST_NGINX_SERVROOT/html/;
	index index.html index.htm;
}	
location /RequestDenied {
	return 412;
}
--- request
GET /child/robots.txt
--- error_code: 200

```

**Comment by lubomudr** [comment]
> Add a test under here: https://github.com/wargio/naxsi/blob/main/unit-tests/tests/05naxsi_advanced_whitelists.t#L414
> 
> Something like this.
> 
> ```
> === WL TEST 8: Whitelists inheritance
> ...
> GET /child/robots.txt
> --- error_code: 200
> ```

Thanks for the example

**Review by wargio — APPROVED** [review]

**Comment by wargio** [comment]
Awesome work! thanks for finding and patching this bug!

**Pull request merged**